### PR TITLE
interpreter: Add varname as positional arg in dep.get_variable()

### PR DESCRIPTION
**Pull request by xclaesse**

### CODE REVIEW

File: mesonbuild/interpreter.py [hunk 1]
```diff
@@ -512,10 +512,16 @@ def partial_dependency_method(self, args, kwargs):
...
     @permittedKwargs({'cmake', 'pkgconfig', 'configtool', 'internal', 'default_value', 'pkgconfig_define'})
     @FeatureNewKwargs('dep.get_variable', '0.54.0', ['internal'])
-    def variable_method(self, args, kwargs):
+    def variable_method(self, args: T.Tuple[T.Optional[str]], kwargs: T.Dict[str, T.Any]) -> str:
+        default_varname = args[0]
+        if default_varname is not None:
+            FeatureNew('0.58.0', 'Positional argument to dep.get_variable()').use(self.subproject)
+            for k in ['cmake', 'pkgconfig', 'configtool', 'internal']:
+                if k not in kwargs:
+                    kwargs[k] = default_varname
```
Comment (dcbaker) [inline mesonbuild/interpreter.py:524]: you could simplify this to
I think you can simplify this to:
```suggestion
            for k in ['cmake', 'pkgconfig', 'configtool', 'internal']:
                kwargs.setdefault(k, default_varname)
```

File: test cases/common/212 dependency get_variable method/meson.build [hunk 2]
```diff
@@ -62,3 +62,4 @@ assert(idep.get_variable(pkgconfig : 'foo', cmake : 'foo', configtool : 'foo',
 
 idep = declare_dependency(variables : ['foo=value'])
 assert(idep.get_variable(internal: 'foo') == 'value')
+assert(idep.get_variable('foo') == 'value')
```
Comment (dcbaker) [inline test cases/common/212 dependency get_variable method/meson.build:65]: I would however like some tests that cover a case like `dep.get_variable('foo', cmake : 'FOO')`, (or pkg-config, same idea) where the specialized variable both is and isn't found, to ensure that the specialized variable has precedence.

**Review by dcbaker — CHANGES_REQUESTED** [review]
I'm still not convinced this is a good idea, it's making something that was explicit implicit, and gives a false sense of "if i just write dep.get_variable('foo') everything will work!" until it silently doesn't.

However, if others like the idea I wont hold it up.

**Comment by eli-schwartz** [comment]
> gives a false sense of "if i just write dep.get_variable('foo') everything will work!"

The alternative is giving the same people the false sense of "if i just write `dep.get_variable(pkgconfig: 'foo')` everything will work!"

It already takes manual work to support every dependency method.

I like this idea, so +1 from me...

File: test cases/common/212 dependency get_variable method/meson.build (hunk 2 above)
Comment (eli-schwartz) [reply test cases/common/212 dependency get_variable method/meson.build:65]: On the previous line we could change to:
```
assert(idep.get_variable('FOO', internal: 'foo') == 'value')
```

**Comment by xclaesse** [comment]
> > gives a false sense of "if i just write dep.get_variable('foo') everything will work!"
> 
> The alternative is giving the same people the false sense of "if i just write `dep.get_variable(pkgconfig: 'foo')` everything will work!"

Exactly, I had to fix that anti-pattern countless times. It's in the top list of reasons why fallback to subproject does not work. In practice the different variable name is really just a CMAKE thing, every project that ship a pkgconfig file will use the same variable name for their internal and pc dep. But often projects will ship pkgconfig OR cmake, not both, which means you rarely actually need to pass different variable names.

File: test cases/common/212 dependency get_variable method/meson.build (hunk 2 above)
Comment (xclaesse) [reply test cases/common/212 dependency get_variable method/meson.build:65]: @eli-schwartz Added that test too, good point.

**Review by eli-schwartz — APPROVED** [review]

File: test cases/common/212 dependency get_variable method/meson.build (hunk 2 above)
Comment (dcbaker) [reply test cases/common/212 dependency get_variable method/meson.build:65]: Please do leave the just `get_variable(internal : 'foo')` we should have a test with and without the default.

Comment (xclaesse) [reply test cases/common/212 dependency get_variable method/meson.build:65]: Sure, I now test the 3 cases.

**Pull request merged**